File: Day323/constructStringFromBT.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public String preOrder(TreeNode root) {
        if (root == null) {
            return "";
        }
        String answer = String.valueOf(root.val);
        if (root.left != null || root.right != null) {
            answer += "(" + preOrder(root.left) + ")";
        }
        if (root.right != null) {
            answer += "(" + preOrder(root.right) + ")";
        }
        return answer;
    }
    public String tree2str(TreeNode root) {
        return preOrder(root);
    }
}

File: Day323/constructStringFromBT.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

 func preOrder(root *TreeNode) string {
    if root == nil {
        return ""
    }
    answer := strconv.FormatInt(int64(root.Val),10)
    if root.Left != nil || root.Right != nil {
        answer += "(" + preOrder(root.Left)+")"
    }
    if root.Right != nil {
        answer += "(" + preOrder(root.Right)+")"
    }
    return answer
}
func tree2str(root *TreeNode) string {
    return preOrder(root)
}